lib/integrations/awsoidc/deploydbservice.go

```diff
@@ -0,0 +1,599 @@
+		}
+	*/
+	serviceConfig.JoinParams = config.JoinParams{
+		TokenName: string(types.JoinMethodIAM) + "-token",
+		Method:    types.JoinMethodIAM,
+	}
+
+	// Disable default services
+	serviceConfig.Auth.EnabledFlag = "no"
+	serviceConfig.Proxy.EnabledFlag = "no"
+	serviceConfig.SSH.EnabledFlag = "no"
+
+	// Enable Discovery Service with a specific Discovery Group.
+	serviceConfig.Discovery.EnabledFlag = "yes"
+	serviceConfig.Discovery.DiscoveryGroup = *req.DiscoveryGroupName
+	// TODO(marco): DiscoveryConfig won't start because it has no matchers.
+	// This should be changed when RFD125 adds the DiscoveryConfig resource.
+
+	serviceConfig.Databases.Service.EnabledFlag = "yes"
+	serviceConfig.Databases.ResourceMatchers = []config.ResourceMatcher{{
+		Labels: req.AgentMatcherLabels,
+	}}
+
+	teleportConfigYAMLBytes, err := yaml.Marshal(serviceConfig)
+	if err != nil {
+		return "", trace.Wrap(err)
+	}
+
+	// This Config is meant to be passed as argument to `teleport start`
+	// Eg, `teleport start --config-string <X>`
+	teleportConfigString := base64.StdEncoding.EncodeToString(teleportConfigYAMLBytes)
+
+	return teleportConfigString, nil
+}
+
+// upsertTask ensures a TaskDefinition with TaskName exists
+func upsertTask(ctx context.Context, clt DeployDBServiceClient, req DeployDBServiceRequest, configB64 string) (*ecsTypes.TaskDefinition, error) {
+	taskAgentContainerImage := fmt.Sprintf(teleportContainerImageFmt, req.TeleportVersion)
+
+	taskDefOut, err := clt.RegisterTaskDefinition(ctx, &ecs.RegisterTaskDefinitionInput{
+		Family: req.TaskName,
+		RequiresCompatibilities: []ecsTypes.Compatibility{
+			ecsTypes.CompatibilityFargate,
+		},
+		Cpu:    &taskCPU,
+		Memory: &taskMem,
+
+		NetworkMode:      ecsTypes.NetworkModeAwsvpc,
+		TaskRoleArn:      &req.TaskRoleARN,
+		ExecutionRoleArn: &req.TaskRoleARN,
+		ContainerDefinitions: []ecsTypes.ContainerDefinition{{
+			Command: []string{
+				"start",
+				"--config-string",
+				configB64,
+			},
+			EntryPoint: []string{"teleport"},
+			Image:      &taskAgentContainerImage,
+			Name:       &taskAgentContainerName,
+			LogConfiguration: &ecsTypes.LogConfiguration{
+				LogDriver: ecsTypes.LogDriverAwslogs,
+				Options: map[string]string{
+					"awslogs-group":         "ecs-" + *req.ClusterName,
+					"awslogs-region":        req.Region,
+					"awslogs-create-group":  "true",
+					"awslogs-stream-prefix": *req.ServiceName + "/" + *req.TaskName,
+				},
+			},
+		}},
+	})
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	return taskDefOut.TaskDefinition, nil
+}
+
+// upsertCluster creates the cluster if it doesn't exist.
+// It will update the cluster if it doesn't have the required capacity provider (FARGATE)
+// It will re-create if its status is INACTIVE.
+// If the cluster status is not ACTIVE, an error is returned.
+// The cluster is returned.
+func upsertCluster(ctx context.Context, clt DeployDBServiceClient, clusterName string) (*ecsTypes.Cluster, error) {
```

Comment:
If we're going to establish a pattern of deploying ECS clusters to inject teleport agents, I think we should have a common configuration for *all* such clusters if possible.  I.e. if I deploy 3 teleport agents to the same region/account I'd want them to all end up in the same cluster with one common set of configuration parameters that don't conflict with one another.  That'll be easier to do if the cluster setup logic is designed to be generic from the beginning.

Would it be possible to make this its own standalone helper rather than being tightly coupled with DB service creation?

Reply:
It's creating a fairly generic Cluster: name, tags, and FARGATE as a capacity provider
So, I don't think we are being coupled with the current creation.

I think we generalize the creation code-wise when we get a new use-case, doing it now, imo, feels premature.